qt/applications/workbench/workbench/plotting/figureinteraction.py

```diff
@@ -254,6 +255,32 @@ def on_mouse_button_release(self, event):
             self.marker_selected_in_double_click_event = None
             self.double_click_event = None
 
+    @staticmethod
+    def on_button_release_legend_bounds_check(event):
+        fig = event.canvas.figure
+
+        for ax in fig.get_axes():
+            legend1 = ax.get_legend()
+            if legend1 is None:
+                continue
+
+            # Figure Size (in pixels)
+            fig_width, fig_height = fig.get_size_inches()
+            dpi = fig.get_dpi()
+            fig_width_px, fig_height_px = fig_width * dpi, fig_height * dpi
+
+            # Legend bounding box (in pixels)
+            bbox_legend = legend1.get_window_extent()
+            x0, y0, x1, y1 = bbox_legend.x0, bbox_legend.y0, bbox_legend.x1, bbox_legend.y1
+
+            outside_window = x1 < 0 or x0 > fig_width_px or y1 < 0 or y0 > fig_height_px
+            # Snap back legend
+            if outside_window:
+                legend1 = ax.legend(loc="upper left")
+                legend_set_draggable(legend1, True)
+                legend1.set_in_layout(False)
```

Comment:
I think here you might want to do something like this: https://github.com/mantidproject/mantid/blob/main/Framework/PythonInterface/mantid/plots/mantidaxes.py#L573-L574

This will guarantee that the legend properties stay the same (position, font etc.). These properties can be set by the user, so the legend location may not always be 'upper left'.